### PR TITLE
Tests for plot_rgb function

### PR DESCRIPTION
**Pull request by lwasser**
This pr has tests for the plot_rgb function.
This addresses issue: 
https://github.com/earthlab/earthpy/issues/178

### CODE REVIEW

File: earthpy/tests/test_plotrgb.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+"""Tests for plot_rgb function"""
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: good call. i just updated the  name of the file and will push it shortly.

File: earthpy/tests/test_plotrgb.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+import numpy as np
+import numpy.ma as ma
+import matplotlib.pyplot as plt
+import pytest
+import rasterio as rio
+from earthpy.plot import plot_rgb
+from earthpy.io import path_to_example
+
+plt.show = lambda: None
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: this is a good question. this seemed to make travis happy and it is what rasterio uses -- see here

https://github.com/mapbox/rasterio/blob/master/tests/test_plot.py 

when i didn't have that i got these travis errors because i think it was trying to plot things but had no frame to render on. i couldn't get it to work otherwise. haven't revisited it for a while but i spent a good bit of time on it and it seemed necessary at the time.Note this wasn't an issue locally it was an issue specific to travis.

File: earthpy/tests/test_plotrgb.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+def test_ax_provided(rgb_image):
+    """Test to ensure the plot works when an explicit axis is provided"""
+    f, ax1 = plt.subplots()
+    f, ax = plot_rgb(rgb_image, ax=ax1)
+
+    rgb_im_shape = rgb_image.transpose([1, 2, 0]).shape
+    the_plot_im_shape = ax.get_images()[0].get_array().shape
+    assert rgb_im_shape == the_plot_im_shape
+    plt.close(f)
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: Yes! do you remember those issues we had with travis? the issue was that travis was failing even tho everything ran locally just fine. i never fully understood WHY this was happening but it seems like a best practice to just clean the slate each time. i fought with this for a while.

File: earthpy/tests/test_plotrgb.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    """Test to ensure the input image has 3 bands to support rgb plotting"""
+    rgb_1band = rgb_image[1]
+
+    with pytest.raises(Exception):
+        plot_rgb(rgb_1band)
+
+
+def test_ax_provided(rgb_image):
+    """Test to ensure the plot works when an explicit axis is provided"""
+    f, ax1 = plt.subplots()
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: ok i think i understand the point of confusion here... i've updated it as you suggest. i didn't understand at first but now i do!!

File: earthpy/tests/test_plotrgb.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    f, ax = plot_rgb(rgb_image, ax=ax1)
+
+    rgb_im_shape = rgb_image.transpose([1, 2, 0]).shape
+    the_plot_im_shape = ax.get_images()[0].get_array().shape
+    assert rgb_im_shape == the_plot_im_shape
+    plt.close(f)
+
+
+def test_ax_not_provided(rgb_image):
+    """Test to ensure that plot_rgb produces a output image on an axis
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: ok so this is the docstring here 

`    """Test to ensure that plot_rgb produces a output image on an axis
    when an axis object is not provided."""`

i don't think we want to add more lines here. so are you suggesting we go through all docstrings and shorten them for the tests? in this case a user wouldn't see them but we will . thoughts?

**Comment by lwasser** [comment]
ok @mbjoseph once travis has run i think this is ready for another review. i've added a new test fun that tests for several parts of a custom plot including figsize, title, rgb and extent. it just seemed simpler to have one plot call then to create a new test and plot for each. have a look and let me know what you think!

File: earthpy/tests/test_plotrgb.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbjoseph) [reply earthpy/tests/test_plotrgb.py]: Cool - thanks!

File: earthpy/tests/test_plotrgb.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbjoseph) [reply earthpy/tests/test_plotrgb.py]: I was asking whether you wanted to have folks adhere to PEP 257 for multi-line docstrings, e.g.: 

```python
def test_ax_not_provided(rgb_image):
    """Test plot_rgb without user-provided axis

    plot_rgb should produce an axis when an axis object is not provided.
    """
```

If not, that's fine. Rasterio seems not to have docstrings for most of their tests, which I think is fine if the test is well-named and easy to understand, but as we start to have more folks contributing tests it could be nice to be able to point to a PEP to provide guidance, particularly if the tests are not well-named and/or hard to follow without extra narrative in a docstring.

File: earthpy/tests/test_plotrgb.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    with rio.open(path_to_example("rmnp-rgb.tif")) as src:
+        rgb = src.read()
+    return rgb
+
+
+def test_1band(rgb_image):
+    """Test to ensure the input image has 3 bands to support rgb plotting"""
+    rgb_1band = rgb_image[1]
+
+    with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (mbjoseph) [reply earthpy/tests/test_plotrgb.py]: I see - in this case it seems like a `ValueError` might be more specific than a generic Exception (https://docs.python.org/3/library/exceptions.html#ValueError):

    Raised when an operation or function receives an argument that has the 
    right type but an inappropriate value

So you could change the `plot_rgb` code to raise a `ValueError` instead of `Exception`, but that's not totally necessary now that you've used the `match` argument. Up to you!

File: earthpy/tests/test_plotrgb.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    rgb_im_shape = rgb_image.transpose([1, 2, 0]).shape
+    the_plot_im_shape = ax.get_images()[0].get_array().shape
+    assert rgb_im_shape == the_plot_im_shape
+    plt.close(f)
+
+
+def test_ax_not_provided(rgb_image):
+    """Test to ensure that plot_rgb produces a output image on an axis
+    when an axis object is not provided."""
+
```
[the remark this answers is not in the document]
Comment (mbjoseph) [reply earthpy/tests/test_plotrgb.py]: Resolved by https://github.com/earthlab/earthpy/issues/207

File: earthpy/tests/test_plotrgb.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: Ok wonderful. I have started a new issue associated with this comment to ensure it doesn't get lost. but let's implement this in a separate PR @mbjoseph   as we just discussed :)

File: earthpy/tests/test_plotrgb.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: Ok @mbjoseph !!! i've updated the ERROR in the function to be a ValueError and then modified the test accordingly. I'm going to see if there is anything else that i missed here... but we might be ready for another review...

File: earthpy/tests/test_plotrgb.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    """Test that if you run clip with a different stretch than the default,
+     you get a different mean value given the change in the range of values
+     stretched."""
+
+    im = rgb_image
+    np.place(im, im > 150, [0])
+
+    f, ax = plot_rgb(im, stretch=True, str_clip=1)
+    mean_val = ax.get_images()[0].get_array().mean()
+    assert mean_val == 110.23530766608626
```
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: hey @mbjoseph I think i generally get what you did above however this doesn't test that the actual stretch is being implemented correctly. it just tests (i think mean_vals is supposed to be stretch_vals maybe?? i'm not sure above but it looks like you are just testing the number of returned values rather than the accuracy of the actual value. 

i wanted to test that the actual stretch worked. Thus i calculated what the mean should be with the val of 1 for the stretch clip. should i just include that code in the function? it will just become a bit more complex that way but i guess then you know where the number comes from. thoughts?

**Comment by lwasser** [comment]
one other note -- this is failing [here](https://travis-ci.org/earthlab/earthpy/builds/487095172#L2684) now but these are not tests that i wrote. it looks like it's test on the data??

**Comment by lwasser** [comment]
Ok @mbjoseph so.... i've made a bunch of changes now to this PR and it's ready for another round:

1. i've pulled the stretch function into a **new helper function**
2. I've added two tests for that helper that explicitly test stretch.

So about testing stretch. it's kind of tricky. i did two tests
1. one that ensures that data without a full range of values is stretched to 0-255.  and
2. one that tests (copying your example above) that varying clips provide varying mean values

I can't think of a good way to test for an ACTUAL value. but that _atleast_ tests that the values vary.  

See what you think. All tests are passing BUT there is that plot warning that i mentioned in my comment above. i haven't found a good solution for that yet. i am totally closing figures each time... 

but that function with the loop creates 10 axis objects that don't seem to be attached to a figure that i can find... i might look into that a bit more .. one option could be to create a loop that creates a fig and ax set of objects... i've done that before. it's only a warning however but... it's still there. perhaps that would close things. BUT the message says there are more than 20 figures open. that seems wrong given that only produces 10 axes.

File: earthpy/tests/test_plotrgb.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lwasser) [reply earthpy/tests/test_plotrgb.py]: ok two new tested are implemented on the new stretch helper function.

**Comment by mbjoseph** [comment]
Hey @lwasser I like the refactored stretch function. Are the new fixtures in `tests/conftest.py` going to be used by other test modules? If not, then it's probably better to define them in the test module that uses them.

**Comment by lwasser** [comment]
hey @mbjoseph i was looking closely at the conftest.py file. I  think this issue of fixture location should be a larger issue that perhaps the interns could address. If you look at the file you will notice many of the fixtures (most even??) are specific to one module. i think the stretch fixture is specific to `plot_rgb` but the others are more generic and i wonder if we could do a larger test and associated fixture cleanup that would address this! so for example basic_geometry() is a fixture only used in test_spatial.

Also i noticed test_spatial is dated given we have been creating most specific modules for sets of functions.

In the meantime i've moved the one fixture that is definitely specific to plot_rgb back. but let me know what you think about a bigger cleanup of fixtures.

**Comment by mbjoseph** [comment]
Great points @lwasser! I opened two issues for us to clean up our conftest.py file (#218) and refactor the tests (#219).

**Comment by lwasser** [comment]
woo hoo.  Thank you @mbjoseph  merging this!!! :)
